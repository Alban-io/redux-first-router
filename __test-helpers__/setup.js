import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
