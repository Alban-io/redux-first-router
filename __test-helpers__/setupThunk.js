import { createStore, applyMiddleware, compose } from 'redux'
import createHistory from 'history/createMemoryHistory'
import connectTypes from '../src/connectTypes'


export default (path = '/', thunk) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk },
    THIRD: { path: '/third/:param' },
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, enhancer, thunk: ssrThunk, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, compose(enhancer, middlewares))

  return { store, thunk: ssrThunk }
}
