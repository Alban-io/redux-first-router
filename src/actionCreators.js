export const INIT = '@@address-bar/INIT' //used internally by middleware and then action is re-written
export const NOT_FOUND = '@@address-bar/NOT_FOUND'


/** EXPORTED INIT ACTION CREATORS: 
 *  for use when ready key is not provided and 
 *  manual control of initialization is desired
*/

export function init(pathname) {
  return {
    type: INIT,
    payload: {
      pathname,
    }
  }
}

export function initThunk() {
  return (dispatch, getState) => {
    let {pathname} = getState().location
    return dispatch(init(pathname))
  }
}


/** PRIMARY/SOLE NAVIGATION ACTION CREATOR 
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'