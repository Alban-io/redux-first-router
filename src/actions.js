export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
