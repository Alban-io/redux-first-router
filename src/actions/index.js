export { default as redirect } from './redirect'
export { default as notFound } from './notFound'
export { default as addRoutes } from './addRoutes'
export { default as changeBasename } from './changeBasename'

export {
  push,
  replace,
  setState,
  jump,
  back,
  next,
  reset
} from './history'
