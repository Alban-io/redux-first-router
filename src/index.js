export {
  default as connectRoutes,
  go,
  back,
} from './connectRoutes'

export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

export { default as actionToPath } from './pure-utils/actionToPath'
export { default as pathToAction } from './pure-utils/pathToAction'

export type {
  RouteString,
  RouteObject,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
 } from './flow-types'

