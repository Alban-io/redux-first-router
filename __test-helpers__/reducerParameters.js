import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
