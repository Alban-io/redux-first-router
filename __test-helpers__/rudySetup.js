import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
