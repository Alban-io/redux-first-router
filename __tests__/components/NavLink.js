import { NOT_FOUND } from '../../src/index'
import { createNavLink, event } from '../../__test-helpers__/createLink'

test('NON-EXACT: show active class', async () => {
  const { tree } = await createNavLink('/first', {
    to: '/first',
    activeClassName: 'active'
  })

  expect(tree).toMatchSnapshot()
})

test('EXACT: DONT show active class', async () => {
  const { tree } = await createNavLink('/second/dog', {
    to: '/second',
    exact: true,
    activeClassName: 'active'
  })

  expect(tree).toMatchSnapshot()
})

test('STRICT: DONT show active class', async () => {
  const { tree } = await createNavLink('/first', {
    to: '/first/',
    strict: true,
    activeClassName: 'active'
  })

  expect(tree).toMatchSnapshot()
})

test('show activeStyle', async () => {
  const { tree } = await createNavLink('/first', {
    to: '/first',
    activeStyle: { color: 'red' }
  })

  expect(tree).toMatchSnapshot()
})

test('combine with existing styles and class', async () => {
  const { tree } = await createNavLink('/first', {
    to: '/first',
    className: 'foo',
    style: { fontSize: 32 },
    activeClassName: 'active',
    activeStyle: { color: 'red' }
  })

  expect(tree).toMatchSnapshot()
})

test('reacts to state changes (onClick)', async () => {
  const { tree, store, component } = await createNavLink('/first', {
    to: '/second/bar',
    activeClassName: 'active'
  })

  expect(tree.props.className).not.toBeDefined()
  expect(tree).toMatchSnapshot() /*? tree */

  // store.dispatch({ type: 'SECOND', payload: { param: 'bar' } })
  await tree.props.onClick(event) // this dispatches above action (obviously)

  const tree2 = component.toJSON() /*? */

  expect(tree2.props.className).toEqual('active')
  expect(tree2).toMatchSnapshot()
})

test('isActive returns true', async () => {
  const { tree, store } = await createNavLink('/first', {
    to: '/first',
    activeClassName: 'active',
    isActive: (match, location) => {
      expect(match).toMatchSnapshot()
      expect(location).toMatchSnapshot()
      return match
    }
  })

  expect(tree).toMatchSnapshot()
})

test('isActive return false', async () => {
  const { tree, component, store } = await createNavLink('/first', {
    to: '/first',
    activeClassName: 'active',
    isActive: (match, location) => {
      expect(match).toMatchSnapshot()
      expect(location).toMatchSnapshot()
      return location.type === 'SECOND'
    }
  })

  expect(tree).toMatchSnapshot()

  await store.dispatch({ type: 'SECOND', payload: { param: 'foo' } })
  expect(component.toJSON()).toMatchSnapshot()
})

it('supports custom HTML tag name', async () => {
  const { tree, store } = await createNavLink('/first', {
    to: '/second',
    activeClassName: 'active',
    tagName: 'div'
  })

  expect(tree).toMatchSnapshot()
})

it('supports custom HTML tag name in active mode', async () => {
  const { tree, store } = await createNavLink('/first', {
    to: '/first',
    activeClassName: 'active-foo',
    tagName: 'div'
  })

  expect(tree).toMatchSnapshot()
})

it('supports custom HTML tag name which is still a link', async () => {
  const { tree, store } = await createNavLink('/first', {
    to: 'somewhere',
    tagName: 'a'
  })

  expect(tree).toMatchSnapshot()
})

test('query params are ommitted', async () => {
  const { tree } = await createNavLink('/first', {
    to: '/first?foo=123',
    activeClassName: 'active'
  })

  expect(tree).toMatchSnapshot()
})
