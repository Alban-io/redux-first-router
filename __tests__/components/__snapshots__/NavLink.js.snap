// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EXACT: DONT show active class 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="/second"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`NON-EXACT: show active class 1`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`STRICT: DONT show active class 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="/first/"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`combine with existing styles and class 1`] = `
<a
  aria-current="true"
  className="foo active"
  href="/first"
  onClick={[Function]}
  style={
    Object {
      "color": "red",
      "fontSize": 32,
    }
  }
/>
`;

exports[`isActive return false 1`] = `
Object {
  "isExact": true,
  "params": Object {},
  "path": "/first",
  "url": "/first",
}
`;

exports[`isActive return false 2`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
  ],
  "hasSSR": undefined,
  "hash": "",
  "index": 0,
  "kind": "load",
  "length": 1,
  "pathname": "/first",
  "payload": Object {},
  "prev": Object {
    "basename": "",
    "entries": Array [],
    "hash": "",
    "index": -1,
    "kind": "",
    "length": 0,
    "pathname": "",
    "payload": Object {},
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "type": "",
    "url": "",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "type": "FIRST",
  "url": "/first",
}
`;

exports[`isActive return false 3`] = `
<a
  aria-current={false}
  className={undefined}
  href="/first"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`isActive return false 4`] = `null`;

exports[`isActive return false 5`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/second/foo",
      "search": "",
      "state": Object {},
      "url": "/second/foo",
    },
  ],
  "hasSSR": undefined,
  "hash": "",
  "index": 1,
  "kind": "push",
  "length": 2,
  "pathname": "/second/foo",
  "payload": Object {
    "param": "foo",
  },
  "prev": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "hasSSR": undefined,
    "hash": "",
    "index": 0,
    "kind": "load",
    "length": 1,
    "pathname": "/first",
    "payload": Object {},
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "type": "FIRST",
    "url": "/first",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "type": "SECOND",
  "url": "/second/foo",
}
`;

exports[`isActive return false 6`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`isActive returns true 1`] = `
Object {
  "isExact": true,
  "params": Object {},
  "path": "/first",
  "url": "/first",
}
`;

exports[`isActive returns true 2`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
  ],
  "hasSSR": undefined,
  "hash": "",
  "index": 0,
  "kind": "load",
  "length": 1,
  "pathname": "/first",
  "payload": Object {},
  "prev": Object {
    "basename": "",
    "entries": Array [],
    "hash": "",
    "index": -1,
    "kind": "",
    "length": 0,
    "pathname": "",
    "payload": Object {},
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "type": "",
    "url": "",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "type": "FIRST",
  "url": "/first",
}
`;

exports[`isActive returns true 3`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`query params are ommitted 1`] = `
<a
  aria-current="true"
  className="active"
  href="/first?foo=123"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`reacts to state changes (onClick) 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="/second/bar"
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`reacts to state changes (onClick) 2`] = `
<a
  aria-current="true"
  className="active"
  href="/second/bar"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`show activeStyle 1`] = `
<a
  aria-current="true"
  className="active"
  href="/first"
  onClick={[Function]}
  style={
    Object {
      "color": "red",
    }
  }
/>
`;

exports[`supports custom HTML tag name 1`] = `
<div
  aria-current={false}
  className={undefined}
  onClick={[Function]}
  style={undefined}
/>
`;

exports[`supports custom HTML tag name in active mode 1`] = `
<div
  aria-current="true"
  className="active-foo"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`supports custom HTML tag name which is still a link 1`] = `
<a
  aria-current={false}
  className={undefined}
  href="somewhere"
  onClick={[Function]}
  style={undefined}
/>
`;
